### PR TITLE
Explain other

### PR DESCRIPTION
**Pull request by ychaker**
# Changelog
- Refactors explain other functionality out of quepid and into splainer-search
- Adds ability to use the explain other functionality for ES as well as Solr

### CODE REVIEW

File: factories/solrSearcherFactory.js [hunk 1]
```diff
@@ -171,6 +178,62 @@
...
+        }
+        return null;
+      };
+
+      var docsWithExplainOther = function(newDocs, fieldSpec, othersExplained) {
+        var docs = [];
+
+        angular.forEach(newDocs, function(doc) {
+          var overridingExplain = getOverridingExplain(doc, fieldSpec, othersExplained);
+          var normalDoc         = normalDocsSvc.createNormalDoc(fieldSpec, doc, overridingExplain);
```
Comment (softwaredoug) [inline factories/solrSearcherFactory.js]: So this service has intentionally avoided depending on normalDocsSvc on purpose, instead producing regular ole Solr/ES docs and expecting the user to normalize them. I would prefer not to have this dependency in the search services, but I need to dig more into what you're doing.

File: services/esSearcherPreprocessorSvc.js [hunk 2]
```diff
@@ -61,6 +62,26 @@ angular.module('o19s.splainer-search')
         }
 
         searcher.queryDsl   = queryDsl;
+      };
+
+      var prepareGetRequest = function (searcher) {
+        searcher.url = searcher.url + '?q=' + searcher.queryText;
+      };
```
Comment (softwaredoug) [inline services/esSearcherPreprocessorSvc.js:69]: Just FYI, you can probably also do this with a POST request (using the query_string or simple_query_string ES query).  I don't know if that helps simplify anything.

File: factories/esSearcherFactory.js [hunk 3]
```diff
@@ -176,10 +189,78 @@
...
       }, function error(msg) {
         activeQueries.count--;
-        thisSearcher.inError = true;
+        self.inError = true;
         return $q.reject(msg);
       });
-    }
+    } // end of search()
+
+    function explainOther (otherQuery, fieldSpec) {
```
Comment (softwaredoug) [inline factories/esSearcherFactory.js]: I would prefer putting the explainOther stuff in an external service that uses the searchers with the normalDocsSvc. Is that possible?

**Comment by softwaredoug** [comment]
I think the changes are ok, but I'm trying to figure out how I would use this in code.
1. I have a searcher, which I've executed a `search` on, so it has a current set of docs
2. I go to `explainOther` -- which is another search action which would generate another set of docs
3. It seems then my docs from 1 are overwritten by 2? 

I would expect step 2 to return a separate set of docs, not overwrite the searchers current docs. But it looks like maybe it overrides the searchers docs (ie the main document set)?

**Comment by softwaredoug** [comment]
I think maybe we're imagining two different things,

I imagined a second service similar to the docFinder service from Quepid. It
1. Created a searcher to pull back "other" docs
2. Asked the original, main searcher for the original query to explain (maybe one doc at a time) some second doc in the context of this searcher
3. Normalize them using the fieldSpec
4. Returned normalized docs

And the searcher has an extra "explain" method that would do something different between Solr and ES and work one doc id at a time.

That's just my imagination, which may not actually map well to reality... :)

**Comment by ychaker** [comment]
@softwaredoug the service you're talking about is not necessary. 

look at https://github.com/o19s/quepid/pull/628/files

using the same code and services that we already have, we can do what you've said without the need to add anything else.

that part can be packaged up in a service of its own, but it's not necessary. 

the reason i did not create a new service, because the new service would duplicate the code in the controller that's in Quepid, and would only normalize the docs which is just:

``` js
var normalizedDocs;
if ( searcher.type === 'solr' ) {
  normalizedDocs = solrExplainExtractorSvc.docsWithExplainOther(searcher.docs, fieldSpec, searcher.othersExplained);
} else if ( searcher.type === 'es' ) {
  normalizedDocs = esExplainExtractorSvc.docsWithExplainOther(searcher.docs, fieldSpec);
}
```

of the whole thing, because the ratingsStore part is not available in `splainer-search`. 

And then the controller in Quepid would still have to do all the other parts. So creating a service to replace an `if` statement is too much.

**Comment by softwaredoug** [comment]
I see, so in what you've built there's two modes of using the searcher 

## Searching

User executes initial "canine" search

```
searcher = createSearcher(..."canine"...)
searcher.search() # returns normal results
```

## Explain-othering

User wonders why "text:law" documents didn't come back for their search.

```
searcher = createSearcher(...origQuery: "canine", "text:law"...) 
searcher.explainOther() # returns normal results
```

Whereas my expectation of the explainOther operation is that it's in the context of an initial search (as you're explaining what happened).

```
theSearcher = createSearcher("canine")
theSearcher.search() # returns normal results

// sometime later, you wonder, why didn't the "law" searches come back
theSearcher.explainOther(..."text:law")
```

I think what used to happen was you created a searcher for "law" then you used the original(?) searcher to explain other? 

Sorry, I don't mean to be argumentative, I was just trying to imagine in Splainer how I'd go about updating to the latest splainer-search to pick up the functionality. And when I looked at your tests to try to figure it out, I only saw one searcher where in my mind there's two search actions, but now I think I get it.

**Comment by softwaredoug** [comment]
Maybe we ought to 
- Add documentation to the README for this (just to clear up any confusion?) What do you think?
- Does this modify any existing APIs/expectations where we need to increment a minor or major version?

**Comment by ychaker** [comment]
@softwaredoug the funny part for me is that i just took the old code that was working in Quepid and replicated the same logic :smile: 

So the method in which I did it wasn't something I made up, it was copy/pasted from Quepid into splainer-search and adjusted for some parts. The way it was done (by you may I add :stuck_out_tongue_winking_eye:), was to take the original query and the new query, perform a search on the new query with the explain other param passed on from the original query. Which makes sense, because technically the new search is not related to the old search.

Think of it this way, i want to explain `text:law`, in relation to the original query, how would I do that in the Solr and ES APIs? 
- I would have to create a new search with the query `q=text:law`, and pass in as one of the params the original query as the explain other param (in Solr you can do that in one call, in ES you have to make multiple calls for each result that comes back from the new query). 
- In this case, what is shared between the old search and the new search? That would be just the settings (API URL, fields to display, etc). But for all intents and purposes, it's a completely new search that shares no context with the existing search. 

So, the way I see it (with my limited knowledge) is that this scenario:

```
theSearcher = createSearcher("canine")
theSearcher.search() # returns normal results

// sometime later, you wonder, why didn't the "law" searches come back
theSearcher.explainOther(..."text:law")
```

works for how the user would be thinking, but to implement that you'd have to think of it this way

```
theSearcher = createSearcher("canine")
theSearcher.search() # returns normal results

// sometime later, you wonder, I know the search "text:law" returns the results I'm looking for
// so why did it not match the search "canine"?
otherSearcher = createSearcher(...) # another searcher with the same "context" as the previous one (same URL, same search engine, ...)
otherSearcher.explainOther("text:law") # returns normal results for "text:law" with an explanation on why it did(n't) match "canine"
```

does that make sense?

**Comment by ychaker** [comment]
For the README, I'll do that now and push it to this branch.

For the version: It doesn't break existing APIs, it adds a new one. so in BREAKING.NEW.BUGFIX, I would increment NEW

**Comment by softwaredoug** [comment]
:+1:

**Pull request merged**